contracts/LightClient.sol

```diff
@@ -144,11 +155,48 @@ contract LightClient is ILightClient, Ownable {
         require(_epochSigners.length == _epochVotingPowers.length, "incorrect length");
 
         uint256 position = _epochPosition(_epochIdx);
+
         curEpochIdx = _epochIdx;
         epochs[position].curEpochVals = _epochSigners;
         epochs[position].curVotingPowers = _epochVotingPowers;
+    }
+
+    function _totalProduceBlock(uint256[] memory produceBlocks) internal pure returns (uint256 total) {
+        for (uint256 i = 0; i < produceBlocks.length; i++) {
+            total += produceBlocks[i];
+        }
+    }
+
+    function _validatorRewardShare(
+        uint256 epochReward,
+        uint256 produceAmount,
+        uint256 totalProduceAmount
+    ) internal pure returns (uint256) {
+        return (epochReward * produceAmount) / totalProduceAmount;
+    }
 
-        // TODO: add rewards to validators
+    function _perEpochReward(address[] memory rewardVals, uint256[] memory produceAmountList) internal {
+        uint256 epochReward = w3qErc20.perEpochReward();
+
+        uint256 totalProduceAmount = _totalProduceBlock(produceAmountList);
+        // Calculate the amount of tokens to reward validator and delegators
+        for (uint256 i = 0; i < rewardVals.length; i++) {
+            address valAddr = rewardVals[i];
+
+            uint256 totalRewardAmount = _validatorRewardShare(epochReward, produceAmountList[i], totalProduceAmount);
+            uint256 valRewardAmount = totalRewardAmount;
+
+            staking.rewardValidator(valAddr, valRewardAmount);
+
+            // reward validator
+            w3qErc20.mint(address(staking), valRewardAmount);
```

Comment:
I guess you could reward all validators once by just `mint(address(staking), epochReward)`?